common/compat.c

```diff
@@ -112,7 +112,28 @@ getprogname (void)
 	if (p != NULL)
 		name = p + 1;
 #elif defined (HAVE_PROGRAM_INVOCATION_SHORT_NAME)
+	const char *p;
 	name = program_invocation_short_name;
+#ifdef __linux__
+	p = strrchr (name ? name : "", '/');
+	if (p != NULL) {
+		/*
+		 * Some programs pack command line arguments into argv[0].
+		 * Check if it is the case by reading /proc/self/exe and extract
+		 * the program name.
+		 *
+		 * Logic borrowed from:
+		 * <https://github.com/mesa3d/mesa/commit/759b94038987bb983398cd4b1d2cb1c8f79817a9>.
+		 */
+		static char *buf;
+
+		if (!buf)
+			buf = realpath ("/proc/self/exe", NULL);
+
+		if (buf && strncmp (buf, name, strlen (buf)) == 0)
```

Comment:
This won't work if name isn't a realpath itself, e.g. a relative path or a symlinked path. Is there a need to cover those cases as well?

Reply:
Indeed; however, it would bring too much complexity in this exceptional path IMO. Let's skip it until we find an actual use-case that require such checks.